### PR TITLE
Adds Iterable eq for checking equality of iterables

### PR DESCRIPTION
**Pull request by ashishkujoy**

### CODE REVIEW

File: kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/eq/Eq.kt [hunk 1]
```diff
@@ -15,15 +17,35 @@ interface Eq<T> {
...
-   shouldShowDataClassDiff(actual, expected) -> DataClassEq.equals(actual as Any, expected as Any)
-   else -> DefaultEq.equals(actual as Any, expected as Any)
+fun <T: Any?> eq(actual: T, expected: T): Throwable? {
+   // if we have null and non null, usually that's a failure, but people can override equals to allow it
+   return when {
+      actual === expected -> null
+      actual == null && expected == null -> null
+      actual == null && expected != null && actual != expected -> actualIsNull(expected)
+      actual != null && expected == null && actual != expected -> expectedIsNull(actual)
+      actual !=null && expected != null -> when {
```
Comment (sksamuel) [inline kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/eq/Eq.kt:27]: What is the null changes for ?

Comment (ashishkujoy) [reply kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/eq/Eq.kt:27]: So this is related to nested equality changes.
Earlier these null checks were in `should` function but there we were throwing error for inequality. While checking each element of Map or Iterables I am now using `eq`,  there I need to have all these null checks else I was getting null pointer exception. So I have moved null checks from `should` to `eq`.

So from:
https://github.com/kotest/kotest/blob/cb23f758ca42669b75962ab642120a762fc2df74/kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/matchers/should.kt#L11-L26

It changes to.
        https://github.com/ashishkujoy/kotest/blob/4065ea2b6908f2063befcdda52d50bd1c22333f1/kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/matchers/should.kt#L14-L16

**Comment by sksamuel** [comment]
I see. Did the error messages change at all ?

**Comment by ashishkujoy** [comment]
Messages are still the same as they were before these changes.

**Comment by sksamuel** [comment]
Ok lgmt merge whenever

**Pull request merged**